### PR TITLE
Clean up optimizer code

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Cleans up some of the `OptimizerConnector` logic to avoid duplicated code.
- Checks if there are extra keys in the `lr_scheduler` dict. Adds a test for this warning.
- Cleans up the `init_optimizers` function so `configure_optimizers` is only called at the end.
- Fixes bug where if configure_optimizers returned `([optimizer], scheduler)`, `trainer.optimizers` became `[optimizer, scheduler]`
- Updates exception message for invalid `configure_optimizers` returns. Adds a test for this exception.
- Updates the default scheduler config to include `strict`.
- Check for extra unsupported keys in the `lr_scheduler` dict. Adds a test for this warning.
- Improves logic to check if there is a monitor when a `ReduceLROnPlateau` scheduler is used. both in the case of `{"optimizer": optimizer, "lr_scheduler": scheduler, "monitor": "your_loss"}` and `{"optimizer": optimizer, "lr_scheduler": {"scheduler": scheduler, "monitor": "your_loss"}}`. Adds tests for these exceptions.
- Updates docs to reflect how we no longer support automatically monitoring `val_loss`.

No issue to tag. Just some stuff I found while coding #3586

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

**Comment by edenlightning** [comment]
Hi @carmocca! Mind rebasing your PR?

**Comment by carmocca** [comment]
> Hi @carmocca! Mind rebasing your PR?

Waiting for #3586. This PR is low priority

**Comment by carmocca** [comment]
@edenlightning @Borda rebased, unblocked and ready to review! :smile:

**Review by awaelchli — APPROVED** [review]
nice

File: tests/trainer/test_optimizers.py [hunk 1]
```diff
@@ -126,55 +126,60 @@ def test_multi_optimizer_with_scheduling_stepping(tmpdir):
...
+    Test exception when lr_scheduler dict has a ReduceLROnPlateau with no monitor
+    """
+    model = EvalModelTemplate()
+    optimizer = torch.optim.Adam(model.parameters())
+    model.configure_optimizers = lambda: {
+        'optimizer': optimizer,
+        'lr_scheduler': {
+            'scheduler': torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer),
+        },
+    }
```
Comment (rohitgr7) [inline tests/trainer/test_optimizers.py:154]: add this to `ComfigureOptimizersPool`?

File: tests/trainer/test_optimizers.py [hunk 2]
```diff
@@ -126,55 +126,60 @@ def test_multi_optimizer_with_scheduling_stepping(tmpdir):
...
-
-    m = '.*ReduceLROnPlateau requires returning a dict from configure_optimizers.*'
-    with pytest.raises(MisconfigurationException, match=m):
+def test_reducelronplateau_with_no_monitor_raises(tmpdir):
+    """
+    Test exception when a ReduceLROnPlateau is used with no monitor
+    """
+    model = EvalModelTemplate()
+    optimizer = torch.optim.Adam(model.parameters())
+    model.configure_optimizers = lambda: ([optimizer], [torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer)])
```
Comment (rohitgr7) [inline tests/trainer/test_optimizers.py:135]:
```suggestion
    model.configure_optimizers = model.configure_optimizers__reduce_lr_on_plateau
```

File: tests/trainer/test_optimizers.py [hunk 3]
```diff
@@ -126,55 +126,60 @@ def test_multi_optimizer_with_scheduling_stepping(tmpdir):
...
-        limit_train_batches=0.2,
-    )
+def test_reducelronplateau_scheduling(tmpdir):
+    model = EvalModelTemplate()
+    optimizer = torch.optim.Adam(model.parameters())
+    model.configure_optimizers = lambda: {
+        'optimizer': optimizer,
+        'lr_scheduler': torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer),
+        'monitor': 'early_stop_on',
+    }
```
Comment (rohitgr7) [inline tests/trainer/test_optimizers.py:167]: add this to `ComfigureOptimizersPool`?

File: tests/trainer/test_optimizers.py [hunk 4]
```diff
@@ -187,46 +192,58 @@ def test_optimizer_return_options():
...
+    assert len(optim) == len(lr_sched) == 1 and len(freq) == 0
+    assert optim[0] == opt_a
+    assert lr_sched[0] == dict(
+        scheduler=scheduler_a, interval='epoch', frequency=1, reduce_on_plateau=False, monitor=None, strict=True
+    )
+
+    # opt tuple of 1 list
+    model.configure_optimizers = lambda: ([opt_a], scheduler_a)
+    optim, lr_sched, freq = trainer.init_optimizers(model)
+    assert len(optim) == len(lr_sched) == 1 and len(freq) == 0
```
Comment (rohitgr7) [inline tests/trainer/test_optimizers.py:221]:
```suggestion
    assert len(optim) == len(lr_sched) == 1
    assert len(freq) == 0
```

File: tests/trainer/test_optimizers.py [hunk 5]
```diff
@@ -187,46 +192,58 @@ def test_optimizer_return_options():
...
-    assert len(optim) == 2 and optim[0] == opt_a and optim[1] == opt_b
-    assert len(lr_sched) == 0 and len(freq) == 0
+    assert optim == [opt_a, opt_b]
+    assert len(lr_sched) == len(freq) == 0
 
     # opt tuple of 2 lists
     model.configure_optimizers = lambda: ([opt_a], [scheduler_a])
     optim, lr_sched, freq = trainer.init_optimizers(model)
-    assert len(optim) == 1 and len(lr_sched) == 1 and len(freq) == 0
+    assert len(optim) == len(lr_sched) == 1 and len(freq) == 0
```
Comment (rohitgr7) [inline tests/trainer/test_optimizers.py:212]:
```suggestion
    assert len(optim) == len(lr_sched) == 1
    assert len(freq) == 0
```

File: tests/trainer/test_optimizers.py [hunk 6]
```diff
@@ -187,46 +192,58 @@ def test_optimizer_return_options():
...
-                               frequency=1, reduce_on_plateau=False)
+    assert lr_sched[0] == dict(
+        scheduler=scheduler_a, interval='epoch', frequency=1, reduce_on_plateau=False, monitor=None, strict=True
+    )
 
     # opt single dictionary
     model.configure_optimizers = lambda: {"optimizer": opt_a, "lr_scheduler": scheduler_a}
     optim, lr_sched, freq = trainer.init_optimizers(model)
-    assert len(optim) == 1 and len(lr_sched) == 1 and len(freq) == 0
+    assert len(optim) == len(lr_sched) == 1 and len(freq) == 0
```
Comment (rohitgr7) [inline tests/trainer/test_optimizers.py:230]:
```suggestion
    assert len(optim) == len(lr_sched) == 1
    assert len(freq) == 0
```

File: pytorch_lightning/trainer/optimizers.py [hunk 7]
```diff
@@ -21,111 +21,107 @@
...
             if isinstance(scheduler, dict):
+                # check provided keys
+                extra_keys = [k for k in scheduler.keys() if k not in default_config.keys()]
+                if extra_keys:
+                    rank_zero_warn(f'Found unsupported keys in the lr scheduler dict: {extra_keys}', RuntimeWarning)
                 if 'scheduler' not in scheduler:
-                    raise ValueError('Lr scheduler should have key `scheduler`',
-                                     ' with item being a lr scheduler')
+                    raise MisconfigurationException(
+                        'The lr scheduler dict must have the key "scheduler" with its item being a lr scheduler'
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/optimizers.py:100]:
```suggestion
                        'The lr scheduler dict must have the key "scheduler" with its item being an lr scheduler'
```

File: pytorch_lightning/trainer/optimizers.py [hunk 8]
```diff
@@ -21,111 +21,107 @@
...
                 lr_schedulers.append({**default_config, **scheduler})
-
             elif isinstance(scheduler, optim.lr_scheduler.ReduceLROnPlateau):
-                lr_schedulers.append({**default_config, 'scheduler': scheduler,
-                                      'reduce_on_plateau': True})
-
+                if monitor is None:
+                    raise MisconfigurationException(
+                        '`configure_optimizers` must include a monitor when a `ReduceLROnPlateau` scheduler is used.'
+                        ' For example: {"optimizer": optimizer, "lr_scheduler": scheduler, "monitor": "your_loss"}'
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/optimizers.py:116]:
```suggestion
                        ' For example: {"optimizer": optimizer, "lr_scheduler": scheduler, "monitor": "metric_to_track"}'
```

File: tests/trainer/test_optimizers.py (hunk 1 above)
Comment (carmocca) [reply tests/trainer/test_optimizers.py:154]: I don't see a point in adding it to the pool if it is not reused in other tests. It just adds extra boilerplate. This way, if the test is updated/removed, you only have to change code in one single place.

File: tests/trainer/test_optimizers.py (hunk 2 above)
Comment (carmocca) [reply tests/trainer/test_optimizers.py:135]: I actually removed it from the pool because it was not being used anywhere else

Comment (rohitgr7) [reply tests/trainer/test_optimizers.py:135]: it might be used sometime later since it's a pretty common one.

File: tests/trainer/test_optimizers.py (hunk 1 above)
Comment (rohitgr7) [reply tests/trainer/test_optimizers.py:154]: hmm. ok :+1:

File: tests/trainer/test_optimizers.py (hunk 2 above)
Comment (carmocca) [reply tests/trainer/test_optimizers.py:135]: Well, the way it was implemented does not work (notice this is a `_raises` test), so I don't think it will be used anywhere else

Comment (rohitgr7) [reply tests/trainer/test_optimizers.py:135]: ok ok yeah. the `monitor` is required.. haha didn't see that. all good :)

**Review by rohitgr7 — APPROVED** [review]
nice!

**Review by SkafteNicki — APPROVED** [review]
LGTM
